Fixes #12988 - Introduce an HTTP/2 frame listener.

sbordet opened the pull request:
Introduced HTTP2Session.FrameListener to receive events about incoming and outgoing frames.

sbordet committed:
Fixes #12988 - Introduce an HTTP/2 frame listener.

Introduced HTTP2Session.FrameListener to receive events about incoming and outgoing frames.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed Parser and Generator now that FrameType's type is not consecutive.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Better fix for Parser and Generator.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
One-off bug.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed StackOverflowError caused by HTTP2SessionContainer being added as a listener to HTTP2Session.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/frames/FrameType.java:
```diff
@@ -29,9 +29,11 @@ public enum FrameType
-    FAILURE(12);
+    // Use high numbers to avoid clashes with newly defined
+    // RFC frames such as ALT-SVC, ORIGIN, etc.
+    PREFACE(125),
+    DISCONNECT(126),
+    FAILURE(127);
```
How about negative numbers for synthetic frames?

gregw reviewed (dismissed):
LGTM, although I think negative IDs for synthetic frames would be better

lorban commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/frames/DisconnectFrame.java:
```diff
@@ -13,6 +13,10 @@
+/**
+ * @deprecated No replacement, never used.
+ */
+@Deprecated(since = "12.0.20")
```
`forRemoval=true`?

lorban commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/generator/NoOpGenerator.java:
```diff
@@ -16,6 +16,10 @@
+/**
+ * @deprecated No replacement, never used.
+ */
+@Deprecated(since = "12.0.20")
```
`forRemoval=true`?

lorban reviewed (dismissed):
Just a couple of nits, up to you what to do with them.

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Introduced HTTP2Session.LifeCycleListener to allow for resource allocation/cleanup of listeners.
It may also serve as a factory of per-session listeners.
Added example of listeners performing low-level PINGs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed ClientConnector.[add|remove]EventListener().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/generator/Generator.java:
```diff
@@ -45,7 +46,7 @@ public Generator(ByteBufferPool bufferPool, boolean useDirectByteBuffers, int ma
         hpackEncoder = new HpackEncoder();
 
-        this.generators = new FrameGenerator[FrameType.values().length];
+        this.generators = new FrameGenerator[FrameType.CONTINUATION.ordinal() + 1];
```
This looks fragile, the array length should be calculated.

sbordet replied:
Not sure how?
The array is filled a few lines below, and any error would be triggered immediately.

lorban replied:
Maybe by adding a `static int lastNonSyntheticIndex()` method to `FrameType`? But I agree that this is minor.

lorban commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java:
```diff
@@ -1150,6 +1210,65 @@ protected void updateLastRemoteStreamId(int streamId)
+
+    private void notifyLifeCycle(BiConsumer<LifeCycleListener, Session> method)
+    {
+        if (lifeCycleListeners.isEmpty())
```
That `isEmpty()` check looks superfluous to me.

lorban commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java:
```diff
@@ -1150,6 +1210,65 @@ protected void updateLastRemoteStreamId(int streamId)
+
+    private void notifyFrame(Consumer<FrameListener> method)
+    {
+        if (frameListeners.isEmpty())
```
That `isEmpty()` check looks superfluous to me.

lorban commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/parser/Parser.java:
```diff
@@ -67,7 +67,7 @@ public Parser(ByteBufferPool bufferPool, int maxHeaderSize, RateControl rateCont
         this.headerParser = new HeaderParser(rateControl == null ? RateControl.NO_RATE_CONTROL : rateControl);
         this.hpackDecoder = new HpackDecoder(maxHeaderSize, this::getBeginNanoTime);
-        this.bodyParsers = new BodyParser[FrameType.values().length];
+        this.bodyParsers = new BodyParser[FrameType.CONTINUATION.ordinal() + 1];
```
The array length should be calculated.

lorban commented on jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/HTTP2ListenersTest.java:
```diff
@@ -0,0 +1,354 @@
+                private void ping(Session session)
+                {
+                    task = httpClient.getScheduler().schedule(() -> close(session), pingTimeout, TimeUnit.MILLISECONDS);
+                    session.ping(new PingFrame(NanoTime.now(), false), Callback.NOOP);
```
Since the nano time isn't used anywhere, maybe it should be replaced with `0L` as I find it a bit confusing.

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban reviewed (dismissed):
The static method in `FrameType` would just be nice to have, but I don't mind the current version.

sbordet committed:
Added documentation.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
